File: sdk/php/src/ContainerService.php
<?php

declare(strict_types=1);

namespace Rexec;

/**
 * Service for managing sandboxes/containers.
 */
class ContainerService
{
    private RexecClient $client;

    public function __construct(RexecClient $client)
    {
        $this->client = $client;
    }

    /**
     * List all containers.
     *
     * Handles production payload `{ containers: [...], count, limit }`
     * (including `containers: null`) and a bare JSON array.
     *
     * @return Container[]
     * @throws RexecException
     */
    public function list(): array
    {
        $response = $this->client->request('GET', '/api/containers');
        $items = self::normalizeList($response);

        return array_map(static fn(array $data) => new Container($data), $items);
    }

    /**
     * Get a container by ID.
     *
     * @throws RexecException
     */
    public function get(string $containerId): Container
    {
        $response = $this->client->request('GET', "/api/containers/{$containerId}");
        if (!is_array($response)) {
            throw new RexecException('Unexpected empty response for get container', 500);
        }

        return new Container($response);
    }

    /**
     * Create a new container.
     *
     * Prefer image aliases such as `ubuntu`, `debian`, or `alpine`
     * (not `ubuntu` on hosted Rexec).
     *
     * @param string $image Image alias or name
     * @param array{name?: string, environment?: array<string, string>, labels?: array<string, string>} $options
     * @throws RexecException
     */
    public function create(string $image, array $options = []): Container
    {
        $body = array_merge(['image' => $image], $options);
        $response = $this->client->request('POST', '/api/containers', $body);
        if (!is_array($response)) {
            throw new RexecException('Unexpected empty response for create container', 500);
        }

        return new Container($response);
    }

    /**
     * Start a container.
     *
     * @throws RexecException
     */
    public function start(string $containerId): void
    {
        $this->client->request('POST', "/api/containers/{$containerId}/start");
    }

    /**
     * Stop a container.
     *
     * @throws RexecException
     */
    public function stop(string $containerId): void
    {
        $this->client->request('POST', "/api/containers/{$containerId}/stop");
    }

    /**
     * Delete a container.
     *
     * @throws RexecException
     */
    public function delete(string $containerId): void
    {
        $this->client->request('DELETE', "/api/containers/{$containerId}");
    }

    /**
     * @return list<array<string, mixed>>
     */
    private static function normalizeList(mixed $response): array
    {
        if ($response === null) {
            return [];
        }
        if (is_array($response) && array_is_list($response)) {
            /** @var list<array<string, mixed>> $response */
            return $response;
        }
        if (is_array($response) && array_key_exists('containers', $response)) {
            $containers = $response['containers'];
            if ($containers === null) {
                return [];
            }
            if (is_array($containers)) {
                /** @var list<array<string, mixed>> $containers */
                return array_values($containers);
            }
        }

        return [];
    }
}

File: sdk/java/src/main/java/io/pipeops/rexec/ContainerService.java
package io.pipeops.rexec;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.Collections;
import java.util.List;

/**
 * Service for managing sandboxes/containers.
 */
public class ContainerService {
    private static final Type CONTAINER_LIST_TYPE = new TypeToken<List<Container>>() {}.getType();

    private final RexecClient client;

    ContainerService(RexecClient client) {
        this.client = client;
    }

    /**
     * List all containers.
     * Handles {@code {containers:[...],count,limit}} (including null list) and a bare array.
     */
    public List<Container> list() throws RexecException {
        String raw = client.requestRaw("GET", "/api/containers", null);
        if (raw == null || raw.isBlank()) {
            return Collections.emptyList();
        }

        JsonElement root = JsonParser.parseString(raw);
        if (root.isJsonArray()) {
            return client.getGson().fromJson(root, CONTAINER_LIST_TYPE);
        }
        if (root.isJsonObject()) {
            JsonObject obj = root.getAsJsonObject();
            if (!obj.has("containers") || obj.get("containers").isJsonNull()) {
                return Collections.emptyList();
            }
            JsonElement containers = obj.get("containers");
            if (containers.isJsonArray()) {
                return client.getGson().fromJson(containers, CONTAINER_LIST_TYPE);
            }
        }
        return Collections.emptyList();
    }

    /**
     * Get a container by ID.
     */
    public Container get(String containerId) throws RexecException {
        return client.request("GET", "/api/containers/" + containerId, null, Container.class);
    }

    /**
     * Create a new container.
     */
    public Container create(CreateContainerRequest request) throws RexecException {
        return client.request("POST", "/api/containers", request, Container.class);
    }

    /**
     * Create a new container with just an image alias.
     */
    public Container create(String image) throws RexecException {
        return create(new CreateContainerRequest(image));
    }

    /**
     * Start a container.
     */
    public void start(String containerId) throws RexecException {
        client.request("POST", "/api/containers/" + containerId + "/start", null, Void.class);
    }

    /**
     * Stop a container.
     */
    public void stop(String containerId) throws RexecException {
        client.request("POST", "/api/containers/" + containerId + "/stop", null, Void.class);
    }

    /**
     * Delete a container.
     */
    public void delete(String containerId) throws RexecException {
        client.request("DELETE", "/api/containers/" + containerId, null, Void.class);
    }
}
